src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt

```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
             sv.analyzer = aa
             aa
         }
+        mv = CoverageBytecodeFilter(
+            skipVisitor.newAdapter(),
+            mv.newAdapter()
+        )
         mv = DeterministicHashCodeTransformer(fileName, className, methodName, mv.newAdapter())
         mv = DeterministicTimeTransformer(mv.newAdapter())
         mv = DeterministicRandomTransformer(fileName, className, methodName, mv.newAdapter())
         return mv
     }
 
 }
+    /**
+     * Filters out chunks of bytecode generated by coverage.
+     *
+     * @param initialAdapter initial adapter to delegate work to when coverage-related bytecode encountered.
+     * @param nextAdapter adapter to delegate calls to when bytecode is not related to coverage.
+     */
+    private class CoverageBytecodeFilter(
+        private val initialAdapter: GeneratorAdapter,
+        nextAdapter: GeneratorAdapter
+    ) : MethodVisitor(ASM_API, nextAdapter) {
+        enum class State {
+            INITIAL,
+            LOADED_HITS_ON_STACK,
```

Comment:
I would suggest some renaming, what do you think?
```
LOADED_HITS_ON_STACK -> INIT
LOADED_HITS_ON_STACK_STATIC -> INIT_FIELD
LOADED_HITS_IN_LOCALS -> IN_LOCAL
PREPARE_ASSIGN_TO_HITS -> ON_STACK
```

Reply:
Should we then start every enum variant with `HITS` prefix?

```
HITS_INIT
HITS_INIT_FIELD
HITS_IN_LOCAL
HITS_ON_STACK
```